File: src/KenpinTool.Prototype/Services/CsvExporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace KenpinTool.Prototype;

public static class CsvExporter
{
    public static void Export(string csvPath, IReadOnlyList<PageItem> pages)
    {
        var dir = Path.GetDirectoryName(csvPath);
        if (!string.IsNullOrWhiteSpace(dir))
        {
            Directory.CreateDirectory(dir);
        }

        var sb = new StringBuilder();
        sb.AppendLine(string.Join(",", new[]
        {
            "page_index",
            "file_name",
            "pdf_page_index",
            "decision_action",
            "decision_ts_utc",
            "exception_reason_code",
            "exception_note",
            "ng_codes",
            "ng_levels",
            "suggested_actions",
            "rework_types",
        }));

        foreach (var page in pages.OrderBy(p => p.Index))
        {
            var decisionAction = page.Decision?.Action.ToString() ?? "";
            var decisionTs = page.Decision?.TimestampUtc.ToString("O") ?? "";
            var excCode = page.Decision?.ExceptionReasonCode ?? "";
            var excNote = page.Decision?.ExceptionNote ?? "";

            var codes = string.Join("|", page.Detections.Select(d => d.Code));
            var levels = string.Join("|", page.Detections.Select(d => Detection.LevelText(d.Level)));
            var suggestedActions = string.Join("|", page.Detections.Select(d => d.SuggestedAction.ToString().ToUpperInvariant()));
            var reworkTypes = string.Join("|", page.Detections.Select(d => d.ReworkType.ToString().ToUpperInvariant()));

            sb.AppendLine(string.Join(",", new[]
            {
                Escape(page.Index.ToString()),
                Escape(page.FileName),
                Escape(page.PdfPageIndex?.ToString() ?? ""),
                Escape(decisionAction),
                Escape(decisionTs),
                Escape(excCode),
                Escape(excNote),
                Escape(codes),
                Escape(levels),
                Escape(suggestedActions),
                Escape(reworkTypes),
            }));
        }

        File.WriteAllText(csvPath, sb.ToString(), new UTF8Encoding(encoderShouldEmitUTF8Identifier: true));
    }

    private static string Escape(string value)
    {
        if (value.Contains('"') || value.Contains(',') || value.Contains('\n') || value.Contains('\r'))
        {
            return "\"" + value.Replace("\"", "\"\"") + "\"";
        }

        return value;
    }
}

File: old_plan/src/kenpintool/services/csv_exporter.py
from __future__ import annotations

import csv
from pathlib import Path

from kenpintool.domain.models import PageItem


class CsvExporter:
    @staticmethod
    def export(path: Path, pages: list[PageItem]) -> None:
        with path.open("w", encoding="utf-8-sig", newline="") as f:
            writer = csv.writer(f)
            writer.writerow(
                [
                    "page_index",
                    "file_name",
                    "pdf_page_index",
                    "decision",
                    "detection_codes",
                    "detection_levels",
                    "suggested_actions",
                    "ai_human_mismatch",
                    "decision_reason",
                    "evidence_count",
                    "override_reason",
                    "ng_codes",
                ]
            )
            for page in pages:
                decision = page.decision.action.value if page.decision else ""
                ng_codes = ",".join(sorted({d.code for d in page.detections}))
                detection_codes = ",".join(d.code for d in page.detections)
                detection_levels = ",".join(d.level.value for d in page.detections)
                suggested_actions = ",".join(
                    d.suggested_action.value if d.suggested_action else ""
                    for d in page.detections
                )
                ai_human_mismatch = bool(page.detections) and decision == "OK"
                decision_reason = (
                    page.decision.exception_reason_code if page.decision else ""
                )
                evidence_count = sum(len(d.evidence) for d in page.detections)
                override_reason = page.decision.override_reason if page.decision else ""
                writer.writerow(
                    [
                        page.index,
                        page.file_name,
                        page.pdf_page_index if page.pdf_page_index is not None else "",
                        decision,
                        detection_codes,
                        detection_levels,
                        suggested_actions,
                        "1" if ai_human_mismatch else "0",
                        decision_reason,
                        evidence_count,
                        override_reason,
                        ng_codes,
                    ]
                )
